components/autofill/src/sync/credit_card/outgoing.rs

```diff
@@ -0,0 +1,101 @@
+use rusqlite::{Transaction, NO_PARAMS};
+use sync_guid::Guid as SyncGuid;
+// use serde_json::*;
+
+pub(super) struct CreditCardsImpl {}
+
+impl ProcessOutgoingRecordImpl for CreditCardsImpl {
+    type Record = InternalCreditCard;
+
+    fn fetch_outgoing_records(
+        &self,
+        tx: &Transaction<'_>,
+        collection_name: String,
+        timestamp: ServerTimestamp,
+    ) -> anyhow::Result<OutgoingChangeset> {
+        // populate outgoing staging table with data records that don't appear in the mirror and data records with a change counter > 0
+
+        // TODO: move this logic to common.rs so it can be share with addresses
+        let mut outgoing = OutgoingChangeset::new(collection_name, timestamp);
+        let data_sql = format!(
+            "SELECT
+                {common_cols},
+                sync_change_counter
+            FROM credit_cards_data
+            WHERE sync_change_counter > 0
+                OR guid NOT IN (
+                    SELECT m.guid
+                    FROM credit_cards_mirror m
+                )",
+            common_cols = CREDIT_CARD_COMMON_COLS
+        );
+
+        let mut outgoing_changes: Vec<Payload> = tx
+            .prepare(&data_sql)?
+            .query_map(NO_PARAMS, |row| {
+                Ok(Payload::from_record(InternalCreditCard::from_row(row)?).unwrap())
+            })?
+            .collect::<std::result::Result<Vec<Payload>, _>>()?;
+
+        let tombstones_sql = "SELECT guid FROM credit_cards_tombstones";
+        let mut outgoing_tombstones: Vec<Payload> = tx
+            .prepare(&tombstones_sql)?
+            .query_map(NO_PARAMS, |row| {
+                Ok(Payload::new_tombstone(SyncGuid::from_string(
+                    row.get("guid")?,
+                )))
+            })?
+            .collect::<std::result::Result<Vec<Payload>, _>>()?;
+
+        outgoing_changes.append(&mut outgoing_tombstones);
+
+        // TODO: add to outgoing staging table
+        // push to outgoing.changes
+        outgoing.changes = outgoing_changes;
+        Ok(outgoing)
+    }
+
+    fn push_synced_items(
+        &self,
+        tx: &Transaction<'_>,
+        _uploaded_at: ServerTimestamp,
+        records_synced: Vec<SyncGuid>,
+    ) -> anyhow::Result<()> {
+        sql_support::each_chunk(&records_synced, |chunk, _| -> anyhow::Result<()> {
+            &tx.execute(
+                &format!(
+                    "UPDATE credit_cards_data
+                    SET sync_change_counter = sync_change_counter - outgoing.sync_change_counter
+                    FROM (SELECT sync_change_counter FROM temp.credit_cards_sync_outgoing_staging LIMIT 1) AS outgoing
+                    WHERE guid IN ({values})
+                        AND credit_cards_data.guid = outgoing.guid",
+                    values = sql_support::repeat_sql_values(chunk.len())
+                ),
+                chunk,
+            )?;
+            Ok(())
+        })?;
+
+        // TODO: upsert to mirror
```

Comment:
This can possibly even just reuse the upsert I just added - `"INSERT OR REPLACE INTO {} (guid, payload)
             SELECT guid, payload FROM temp.{}"` - it's just that the table names are different for outgoing than incoming.